### PR TITLE
Problem: validators are not sorted by checksum address

### LINKED ISSUE #427: Problem on ordering valset
When creating a valset tx, we are sorting valset (or signer set) using string sorting in golang  (On the ethereum address in case of same validators having the same power).

```
func EthereumAddrLessThan(e, o string) bool {
	return bytes.Compare([]byte(e)[:], []byte(o)[:]) == -1
}
``` 

This will result 

```
0x8D6d2605550c6AFbf0669E340d9Eb90372A06a9d < 0x8c474b92c315FEdBef9b1DaFEe0D5DB45575Ed5C
```

However in rust, the address is deserialized to an H160 (case insensitive), and when checking is correct, it will expect 

```
0x8c474b92c315fedbef9b1dafee0d5db45575ed5c < 0x8d6d2605550c6afbf0669e340d9eb90372a06a9d
```

This will result frequent warning such as and may block the valset updating?

```
Jul 04 14:43:08 ip-10-202-3-42 gorc[174678]: 2022-07-04T14:43:08.478362Z  WARN gravity_utils::types::ethereum_events: Someone submitted an unsorted validator set, this means all updates will fail until someone feeds in this unsorted value by hand 
[
ValsetMember { power: 1297573200, eth_address: Some(0x5044c263c810554dd4fd33fa5f303c412b5c350b) },
ValsetMember { power: 1297573200, eth_address: Some(0x5be487aec16780e7d38beb3ca56c90b1b562a53b) },
ValsetMember { power: 1297573200, eth_address: Some(0xa198923ddac5dee8df1009c85315781c7fcb7bc1) },
ValsetMember { power: 77854392, eth_address: Some(0xef450b7a2c5394ddf923b17cf7be08a30ff94c0d) },
ValsetMember { power: 64878660, eth_address: Some(0x0712ec6f893d306f0fe1a6de062ce7cd8a6556ae) },
ValsetMember { power: 64878660, eth_address: Some(0x66d74eca51ea8f1df7510454002191c6769915a8) },
ValsetMember { power: 64878660, eth_address: Some(0x8d6d2605550c6afbf0669e340d9eb90372a06a9d) },
ValsetMember { power: 64878660, eth_address: Some(0x8c474b92c315fedbef9b1dafee0d5db45575ed5c) },
ValsetMember { power: 64878660, eth_address: Some(0xee4b98cb5f57e2bf1dade9e794f7c92e500b0d2b) }
]
instead of 
[
ValsetMember { power: 1297573200, eth_address: Some(0x5044c263c810554dd4fd33fa5f303c412b5c350b) },
ValsetMember { power: 1297573200, eth_address: Some(0x5be487aec16780e7d38beb3ca56c90b1b562a53b) },
ValsetMember { power: 1297573200, eth_address: Some(0xa198923ddac5dee8df1009c85315781c7fcb7bc1) },
ValsetMember { power: 77854392, eth_address: Some(0xef450b7a2c5394ddf923b17cf7be08a30ff94c0d) },
ValsetMember { power: 64878660, eth_address: Some(0x0712ec6f893d306f0fe1a6de062ce7cd8a6556ae) },
ValsetMember { power: 64878660, eth_address: Some(0x66d74eca51ea8f1df7510454002191c6769915a8) },
ValsetMember { power: 64878660, eth_address: Some(0x8c474b92c315fedbef9b1dafee0d5db45575ed5c) },
ValsetMember { power: 64878660, eth_address: Some(0x8d6d2605550c6afbf0669e340d9eb90372a06a9d) },
ValsetMember { power: 64878660, eth_address: Some(0xee4b98cb5f57e2bf1dade9e794f7c92e500b0d2b) }
]
```

### PR DESCRIPTION
**Pull request by thomas-nguy**
Fix https://github.com/PeggyJV/gravity-bridge/issues/427

### CODE REVIEW

File: orchestrator/gravity_utils/Cargo.toml [hunk 1]
```diff
@@ -11,6 +11,7 @@ gravity_abi = { path = "../gravity_abi" }
 gravity_proto = { path = "../gravity_proto/" }
 cosmos-sdk-proto = "0.6.3"
 deep_space = { git = "https://github.com/iqlusioninc/deep_space/", branch = "master" }
+eth_checksum = "0.1.2"
```
Comment (devashishdxt) [inline orchestrator/gravity_utils/Cargo.toml]: Will it be a good idea to vendor in the `eth_checksum` code in the repo? It's only a single function ([here](https://github.com/miguelmota/rust-eth-checksum/blob/master/src/lib.rs)). Will reduce one dependency.

Comment (devashishdxt) [reply orchestrator/gravity_utils/Cargo.toml]: And we already have `sha3` as a dependency.

Comment (thomas-nguy) [reply orchestrator/gravity_utils/Cargo.toml]: good call, I have removed the dependency

File: orchestrator/gravity_utils/src/types/checksum.rs [hunk 2]
```diff
@@ -0,0 +1,29 @@
+use ethers::utils::hex;
+use sha3::{Digest, Keccak256};
+
+pub fn checksum(address: &str) -> String {
```
Comment (tomtau) [inline orchestrator/gravity_utils/src/types/checksum.rs:6]: some ref doc that describes how the checksum is computed or attribution to the original code where it comes from if it was copy-pasted from somewhere?

File: orchestrator/gravity_utils/src/types/valsets.rs [hunk 3]
```diff
@@ -339,7 +341,13 @@ impl Ord for ValsetMember {
...
             self.power.cmp(&other.power)
         } else {
-            self.eth_address.cmp(&other.eth_address).reverse()
+            // Since gravity-bridge and ethers use checksum addresses
+            // we need to convert in order to do the comparison.
+            let checksum_addr = checksum::checksum(
+                &self.eth_address.unwrap_or_default().to_hex());
+            let checksum_other = checksum::checksum(
+                &other.eth_address.unwrap_or_default().to_hex());
+            checksum_addr.cmp(&checksum_other).reverse()
```
Comment (tomtau) [inline orchestrator/gravity_utils/src/types/valsets.rs]: and ethers doesn't implement comparison on addresses?

File: orchestrator/gravity_utils/src/types/checksum.rs (hunk 2 above)
Comment (thomas-nguy) [reply orchestrator/gravity_utils/src/types/checksum.rs:6]: added

File: orchestrator/gravity_utils/src/types/valsets.rs (hunk 3 above)
Comment (thomas-nguy) [reply orchestrator/gravity_utils/src/types/valsets.rs]: you mean an hex address (hash) or checksummed address?

Comment (tomtau) [reply orchestrator/gravity_utils/src/types/valsets.rs]: checksummed addressed -- the comments says ethers uses checksummed addresses, so why to export to hex and compute a checksum address if the original was a checksummed address?
or is the comment incorrect that gravity bridge uses checksummed addresses, but ethers doesn't?

Comment (thomas-nguy) [reply orchestrator/gravity_utils/src/types/valsets.rs]: yeah basically the ethereum client uses checksummed addresses when reading and relaying the event to the chain.
The gravity bridge also use the checksummed address (well technically string)

The problem is that the tendermint client deserialized the value using Address=H160 which is not checksummed

Comment (thomas-nguy) [reply orchestrator/gravity_utils/src/types/valsets.rs]: I removed the comment on ethers which may be misleading

**Review by tomtau — DISMISSED** [review]

**Pull request closed without merging**